packages/vite/src/node/optimizer/index.ts

```diff
@@ -496,22 +498,152 @@ export async function runOptimizeDeps(
 
   const qualifiedIds = Object.keys(depsInfo)
 
-  const processingResult: DepOptimizationResult = {
+  let cleaned = false
+  const cleanUp = () => {
+    if (!cleaned) {
+      cleaned = true
+      fs.rmSync(processingCacheDir, { recursive: true, force: true })
```

Comment:
as long as processingCacheDir isn't unique to the process, an in memory `cleaned` flag isn't sufficient to guard against parallel processes.

This might be migitated in a follow-up PR that tackles parallel runs.

Things to consider:
- two or more builds might be called simultaneously
- code changes could have happened in between
- the last one started should win, but optimize being async it could be the first to finish

consequences:
- either keep output completely in memory or in a separate temp dir per process
- use a unique token or another approach to ensure that no other run has been started after your own
- optionally check that token during execution (would allow the first started build to fail immediately if you started a new one)

The whole parallel thing is a bit theoretical as people are not really supposed to do that i think. Parallel building for ssr/web would write to different output directories so that is fine, but two parallel client builds or two parallel dev-servers on the same workspace uh-oh.

Reply:
The `cleaned` flag is only guarding against calling `cancel` twice and be able to have a logical error when `commit` is called after `cancel`. As you explain, in a future PR I'll check using different names for each temp dir that I think it is the proper way to handle parallel processes here.

Reply:
:+1: